// @inrupt/solid-client-errors/dist/index.mjs
export { InruptClientError } from './clientError.mjs';
export { ClientHttpError } from './http/httpError.mjs';
export { handleErrorResponse } from './http/handleErrorResponse.mjs';
export { DEFAULT_TYPE, PROBLEM_DETAILS_MIME, hasProblemDetails } from './http/problemDetails.mjs';
export { hasErrorResponse } from './http/errorResponse.mjs';
export { BAD_REQUEST_STATUS, BadRequestError } from './http/wellKnown/badRequestError.mjs';
export { CONFLICT_STATUS, ConflictError } from './http/wellKnown/conflictError.mjs';
export { FORBIDDEN_STATUS, ForbiddenError } from './http/wellKnown/forbiddenError.mjs';
export { GONE_STATUS, GoneError } from './http/wellKnown/goneError.mjs';
export { INTERNAL_SERVER_ERROR_STATUS, InternalServerError } from './http/wellKnown/internalServerError.mjs';
export { METHOD_NOT_ALLOWED_STATUS, MethodNotAllowedError } from './http/wellKnown/methodNotAllowedError.mjs';
export { NOT_ACCEPTABLE_STATUS, NotAcceptableError } from './http/wellKnown/notAcceptableError.mjs';
export { NOT_FOUND_STATUS, NotFoundError } from './http/wellKnown/notFoundError.mjs';
export { PRECONDITION_FAILED_STATUS, PreconditionFailedError } from './http/wellKnown/preconditionFailedError.mjs';
export { TOO_MANY_REQUESTS_STATUS, TooManyRequestsError } from './http/wellKnown/tooManyRequestsError.mjs';
export { UNAUTHORIZED_STATUS, UnauthorizedError } from './http/wellKnown/unauthorizedError.mjs';
export { UNSUPPORTED_MEDIA_TYPE_STATUS, UnsupportedMediaTypeError } from './http/wellKnown/unsupportedMediaTypeError.mjs';
